lib/upgrade/migrator/migrate.go

```diff
@@ -657,53 +660,110 @@ func (m *migrationHelper) queryAndConvertV1ToV3Nodes(data *MigrationData) error
+		log.Debugf("ClusterInformation contained CalicoVersion '%s' and indicates migration is not needed", ci.Spec.CalicoVersion)
 		return false, nil
 	} else if _, ok := err.(cerrors.ErrorResourceDoesNotExist); !ok {
-		// The error indicates a problem with accessing the resource
-		return false, err
+		// The error indicates a problem with accessing the resource.
+		return false, fmt.Errorf("unable to query ClusterInformation to determine Calico version: %v", err)
 	}
-	// The resource does not exist from the clientv3 so we need to check the
-	// clientv1 version.
 
-	// Grab the version from the clientv1
+	// The resource does not exist from the clientv3 so we need to check the
+	// clientv1 version.  Grab the version from the clientv1.
 	v, err := m.getV1ClusterVersion()
 	if err != nil {
 		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); ok {
 			log.Debugf("CalicoVersion does exist in the v1 or v3 data, no migration needed")
 			// The resource does not exist in the clientv1 (or in the clientv3)
 			// so no migration is needed because it seems that this is an
-			// unitialized datastore.
+			// uninitialized datastore.
 			return false, nil
 		}
 
 		// The error indicates a problem accessing the resource.
-		return false, err
+		return false, fmt.Errorf("unable to query global Felix configuration to determine Calico version: %v", err)
 	}
 
-	// Migrate only if it is possible to migrate from the current version
+	// Migrate only if it is possible to migrate from the current version.
 	if yes, err := versionRequiresMigration(v); err != nil {
-		log.Errorf("Unable to migrate version %s: %v", v, err)
-		return false, fmt.Errorf("Unable to migrate version %s: %v", v, err)
+		// Hit an error parsing the version, or we determined that the version is not
+		// valid to migrate from.
+		log.Errorf("Unable to migrate data from version '%s': %v", v, err)
+		return false, fmt.Errorf("unable to migrate data from version '%s': %v", v, err)
 	} else if !yes {
-		log.Errorf("Migration to v3 requires a base of Calico v2.6.4+, currently at %s", v)
-		return false, errors.New(fmt.Sprintf("Migration to v3 requires a base of Calico v2.6.4+, currently at %s", v))
-	}
-	log.Debugf("GlobalConfig contained CalicoVersion %s and indicates migration is needed", v)
+		// Version indicates that migration is not required - however, we should never have a
+		// v3.0+ version in the v1 API data which suggests a modified/hacked set up which we
+		// cannot migrate from.
+		log.Errorf("Unexpected version in the global Felix configuration, currently at %s", v)
+		return false, fmt.Errorf("unexpected Calico version '%s': migration to v3 should be from a tagged "+
+			"release of Calico v%s+", v, minUpgradeVersion)
+	}
+	log.Debugf("GlobalConfig contained CalicoVersion '%s' and indicates migration is needed", v)
 	return true, nil
 }
 
+// CanMigrate checks version information and reports if migration is possible (nil error).
```

Comment:
It was easier to just include this in the cherry-pick ... it won't be used unless you do a new v3.0 specific release of calico-upgrade.